packages/swingset-liveslots/test/virtual-objects/test-vo-real-gc.js

```diff
@@ -0,0 +1,74 @@
+/* global WeakRef */
+import test from 'ava';
+import '@endo/init/debug.js';
+
+import { Far } from '@endo/marshal';
+import { kser, kunser } from '../kmarshal.js';
+import { setupTestLiveslots } from '../liveslots-helpers.js';
+
+test('VO holds presence', async t => {
+  let monitor;
+
+  const initData = { begin: 'ning' };
+  const initStateData = { begin: kser(initData.begin) };
+
+  function buildRootObject(vatPowers) {
+    const { VatData } = vatPowers;
+    const { defineKind } = VatData;
+    const init = () => initData;
+    const makeThing = defineKind('thing', init, {
+      // eslint-disable-next-line no-unused-vars
+      ping: ({ state }) => 0,
+    });
+    const root = Far('root', {
+      make: () => {
+        const thing = makeThing();
+        monitor = new WeakRef(thing);
+        return thing;
+      },
+      ping: thing => {
+        monitor = new WeakRef(thing);
+        return thing.ping();
+      },
+    });
+    return root;
+  }
+
+  const tl = await setupTestLiveslots(t, buildRootObject, 'vatA');
+  const { v, dispatchMessageSuccessfully } = tl;
+  v.log.length = 0;
+
+  // creating the VO will create the initial state data, but does not
+  // require creation of a "context" nor the "state" accessor object
+  const res = await dispatchMessageSuccessfully('make');
+  const thingKSlot = kunser(res);
+  const kref = thingKSlot.getKref(); // should be o+v10/1
+  const vkey = `vom.${kref}`;
+  const isGet = l => l.type === 'vatstoreGet';
+  const isSet = l => l.type === 'vatstoreSet';
+  const getReads = log => log.filter(l => isGet(l) && l.key === vkey);
+  const getWrites = log => log.filter(l => isSet(l) && l.key === vkey);
+  const getValues = log => getWrites(log).map(l => l.value);
+
+  // the initial data should be written by end-of-crank
+  t.deepEqual(getReads(v.log), []);
+  t.deepEqual(JSON.parse(getValues(v.log)[0]), initStateData);
```

Comment:
No, actually, what gets written into the vatstore is `JSON.stringify({ propA: capdataA, propB: capdataB ..})`, since we want to have one vatstore key for the whole virtual object, but we don't want to have to re-serialize the values of all properties any time we change a single property. The inner serialization uses marshalling (it holds slots/vrefs), but the outer serialization just needs to aggregate a number of capdata structures (body+slots) into a single vatstore string. The inner serialization can be examined by the test program with `kser` (note that `initStateData =` uses kser), but the VOM uses JSON for the outer serialization, so the test program must too.

In the long run, I suspect we'll enhance vatstore to allow a two-level key (taking stronger advantage of SQLite), and make a more specialized set of syscalls to manage virtual-object data. Then we can `vostore.set(baseref, propname, capdata)`, instead of doing this sort of aggregation in userspace. (I'm also thinking we could reveal slots to the DB, and get its help with finding reference edges and performing GC).